src/solvers/refinement/string_constraint_generator_format.cpp

```diff
@@ -0,0 +1,424 @@
+/// \param ref_type: a type  for refined string type
+/// \return String expression representing the output of String.format.
+string_exprt string_constraint_generatort::add_axioms_for_format(
+  const std::string &s,
+  const std::vector<exprt> &args,
+  const refined_string_typet &ref_type)
+{
+  const std::vector<format_elementt> format_strings=parse_format_string(s);
+  std::vector<string_exprt> intermediary_strings;
+  std::size_t arg_count=0;
+
+  for(const format_elementt &fe : format_strings)
+    if(fe.is_format_specifier())
+    {
+      // TODO we shouldn't increment arg_count for constants
+      const format_specifiert &fs=fe.get_format_specifier();
+      struct_exprt arg;
+      if(fs.conversion!=format_specifiert::PERCENT_SIGN &&
+         fs.conversion!=format_specifiert::LINE_SEPARATOR)
+        arg=to_struct_expr(args[arg_count++]);
+      intermediary_strings.push_back(
+        add_axioms_for_format_specifier(fs, arg, ref_type));
+    }
+    else
+      intermediary_strings.push_back(
+        add_axioms_for_constant(
+          fe.get_format_text().get_content(), ref_type));
+
+  if(intermediary_strings.empty())
+    return empty_string(ref_type);
+
+  string_exprt str=intermediary_strings[0];
+  for(std::size_t i=1; i<intermediary_strings.size(); i++)
+    str=add_axioms_for_concat(str, intermediary_strings[i]);
+  return str;
+}
+
+/// Construct a string from a constant array.
+/// \param arr: an array expression containing only constants
+/// \param length: an unsigned value representing the length of the array
+/// \return String of length `length` represented by the array assuming each
+///   field in `arr` represents a character.
+std::string string_constraint_generatort::string_of_constant_array(
+  const array_exprt &arr, unsigned length)
+{
+  std::ostringstream result;
+
+  for(size_t i=0; i<arr.operands().size() && i<length; i++)
+  {
+    // TODO: factorize with utf16_little_endian_to_ascii
+    unsigned c;
+    exprt arr_i=arr.operands()[i];
+    if(arr_i.id()!=ID_constant)
+      throw "non constant format specifier";
+    to_unsigned_integer(to_constant_expr(arr_i), c);
+    if(c<=255 && c>=32)
+      result << (unsigned char) c;
+    else
+    {
+      result << "\\u" << std::hex << std::setw(4) << std::setfill('0')
+             << (unsigned int) c;
+    }
+  }
+  return result.str();
+}
+
+/// Add axioms to specify the Java String.format function.
+///
+/// TODO: This is correct only if the first argument (ie the format string) is
+/// constant or does not contain format specifiers.
+/// \param f: a function application
+/// \return A string expression representing the return value of the
+///   String.format function on the given arguments, assuming the first argument
+///   in the function application is a constant. Otherwise the first argument is
+///   returned.
+string_exprt string_constraint_generatort::add_axioms_for_format(
+  const function_application_exprt &f)
+{
```

Comment:
Can we add assertions (preconditions) on what we expect `f` to be? For instance, I think we should check that `f.arguments()` is not empty, but there might be other things.